util/misc.py

```diff
@@ -428,6 +428,40 @@ def available_cpu_count():
 
     return multiprocessing.cpu_count()
 
+def sanitize_thread_count(threads=None, tool_max_cores_value=available_cpu_count):
+    ''' Given a user specified thread count, this function will:
+        - ensure that 1 <= threads <= available_cpu_count()
+        - interpret None values to mean max available cpus
+            unless PYTEST_XDIST_WORKER_COUNT is defined as an environment
+            variable, in which case we always return 1
+        - allow for various, tool-specific ways of specifying
+            max available cores (tool_max_value)
+        tool_max_cores_value can be one of:
+            available_cpu_count - this function will return available_cpu_count()
+            any other value - this function will return that value.
+            some commonly used values for tools are -1, 0, and None.
+    '''
+    if 'PYTEST_XDIST_WORKER_COUNT' in os.environ:
+        threads = 1
+
+    max_cores = available_cpu_count()
+
+    if threads is None:
+        threads = max_cores
+
+    if type(threads) != int:
+        raise Exception()
```

Comment:
Maybe just an assert here instead of bare exception

Reply:
My understanding is that python `assert` works well in unit tests and in most scenarios, but if anyone ever calls our code with `python -O` it's all blindly ignored.

Though I guess this is specific example is one of the more defensive coding scenarios where it's not strictly necessary (because everyone should be passing in ints in theory), so maybe it's okay for this to be skipped in some cases.